[GStreamer][MSE][CMake] make configurable video decoder limits

pgorszkowski-igalia opened the pull request:
#### 8e1c25cdff34ac6bd154d6abbf4d970c015e22b9
<pre>
[GStreamer][MSE][CMake] make configurable video decoder limits
<a href="https://bugs.webkit.org/show_bug.cgi?id=245660">https://bugs.webkit.org/show_bug.cgi?id=245660</a>

Reviewed by Xabier Rodriguez-Calvar.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(videoDecoderLimitsDefaults):
(WebCore::MediaPlayerPrivateGStreamerMSE::MediaPlayerPrivateGStreamerMSE):
(WebCore::MediaPlayerPrivateGStreamerMSE::supportsType):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/255380@main">https://commits.webkit.org/255380@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a1bf0941d90a4c1ea3164e47d5701948cd05694

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91767 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101461 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161595 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1016 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29545 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84108 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97798 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97425 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/608 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78378 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27573 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82527 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82157 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70616 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83186 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35882 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16181 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78234 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33637 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17282 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27003 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3730 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37484 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40007 "Found 2 new test failures: imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.html?h264_annexb, imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.worker.html?h264_annexb (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80849 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39385 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36400 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17710 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -56,20 +56,46 @@
 #include <wtf/text/AtomStringHash.h>
 #include <wtf/text/StringToIntegerConversion.h>
 
+static void videoDecoderLimitsDefaults(unsigned& mediaMaxWidth, unsigned& mediaMaxHeight, unsigned& mediaMaxFrameRate)
```
Better return a boolean indicating if parsing has been successful (true) or not (false).

calvaris replied:
Even better to return a `std::optional<std::tuple<unsigned, unsigned, unsigned>>`.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -56,20 +56,46 @@
+
+    Vector<String> entries;
+
+    // extract frame rate part from the VIDEO_DECODING_LIMIT: WIDTHxHEIGHT@FRAMERATE
```
I don't sympathize a lot with this WebKit style rule, but... comments must be formal sentences, starting with capital leter and finishing with a dot (.).

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -56,20 +56,46 @@
+    });
+
+    if (entries.size() != 2)
+        return;
```
Return false for all these early returns.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -56,20 +56,46 @@
+    if (entries.size() != 2)
+        return;
+
+    mediaMaxWidth = parseIntegerAllowingTrailingJunk<unsigned>(entries[0]).value_or(0);
```
Instead of `value_or()`, use `!has_value()` to early return false when the optional value isn't set.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -56,20 +56,46 @@
+
+    mediaMaxWidth = parseIntegerAllowingTrailingJunk<unsigned>(entries[0]).value_or(0);
+    mediaMaxHeight = parseIntegerAllowingTrailingJunk<unsigned>(entries[1]).value_or(0);
+#endif
```
Everything has been successful until here, return true before the `#endif`. Return false in an `#else`, for those cases where `VIDEO_DECODING_LIMIT` isn't defined.

pgorszkowski-igalia replied:
Just to be sure your intention here: if we return false in #else (VIDEO_DECODING_LIMIT isn't defined) then we will get error message in that case too

eocanha replied:
Hmmm... your're right. Maybe in that case it makes sense to return true and just leave the default initial values (zero). It doesn't make sense to bug the user in the logs about an error that isn't actually an error.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
```diff
@@ -116,6 +116,10 @@ class MediaPlayerPrivateGStreamerMSE : public MediaPlayerPrivateGStreamer {
+
+    static unsigned mediaMaxWidth;
+    static unsigned mediaMaxHeight;
+    static unsigned mediaMaxFrameRate;
```
Add another boolean static variable to flag if parsing has been attempted (even if it was unsuccessful).

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -88,6 +114,10 @@ GST_DEBUG_CATEGORY(webkit_mse_debug);
+unsigned MediaPlayerPrivateGStreamerMSE::mediaMaxWidth = 0;
+unsigned MediaPlayerPrivateGStreamerMSE::mediaMaxHeight = 0;
+unsigned MediaPlayerPrivateGStreamerMSE::mediaMaxFrameRate = 0;
+
```
Add another boolean static variable to flag if parsing has been attempted (even if it was unsuccessful).

ceyusa commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -401,11 +432,11 @@ MediaPlayer::SupportsType MediaPlayerPrivateGStreamerMSE::supportsType(const Med
 
-    if (width > MEDIA_MAX_WIDTH || height > MEDIA_MAX_HEIGHT)
+    if (width > mediaMaxWidth || height > mediaMaxHeight)
         return result;
```
If I understand correctly, if VIDEO_DECODING_LIMIT cannot be parsed, IsNotSupported will be always returned. Right?

Wouldn't be better define some "safe" default values?

pgorszkowski-igalia replied:
Ups, you are right, will fix that.

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -123,6 +153,7 @@ MediaPlayerPrivateGStreamerMSE::MediaPlayerPrivateGStreamerMSE(MediaPlayer* play
     : MediaPlayerPrivateGStreamer(player)
 {
     GST_TRACE("creating the player (%p)", this);
+    videoDecoderLimitsDefaults(mediaMaxWidth, mediaMaxHeight, mediaMaxFrameRate);
```
Call `videoDecoderLimitsDefaults()` only if parsing hasn't been attempted before. Log an error if parsing was unsuccessful. Set the "parsing has been attempted" flag to true inconditionally.

pgorszkowski-igalia replied:
this function has to be called inside MediaPlayerPrivateGStreamerMSE::supportsType because it is a static function which is called before creation of MediaPlayerPrivateGStreamerMSE

eocanha replied:
Yes, but supportsType() can be called multiple times (to check multiple type/resolution variants). There's no need to parse the limit defaults every time. That's why I was suggesting the "parsing has been attempted" static flag, to parse the values specified at build time (ie: never changing at runtime) only once.

pgorszkowski-igalia replied:
Ah, I was not precisely in my comment, my intention was to signaling that I will move calling videoDecoderLimitsDefaults to supportsType with all your suggestions (which are perfectly fine as you described)

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -401,11 +432,11 @@ MediaPlayer::SupportsType MediaPlayerPrivateGStreamerMSE::supportsType(const Med
         height = 0;
 
-    if (width > MEDIA_MAX_WIDTH || height > MEDIA_MAX_HEIGHT)
+    if (width > mediaMaxWidth || height > mediaMaxHeight)
```
If mediaMaxWidth is zero it means that there's no width limit. Same for height. The if should remain something like this:
```
if ((mediaMaxWidth && width > mediaMaxWidth) || (mediaMaxHeight && height > mediaMaxHeight))
```

eocanha commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -401,11 +432,11 @@ MediaPlayer::SupportsType MediaPlayerPrivateGStreamerMSE::supportsType(const Med
 
     float framerate = parameters.type.parameter("framerate"_s).toFloat(&ok);
-    if (ok && framerate > MEDIA_MAX_FRAMERATE)
+    if (ok && framerate > mediaMaxFrameRate)
```
Same as above: mediaMaxFramerate being zero should mean "no limit". Add a condition to apply the comparison only if there's a valid limit.

pgorszkowski-igalia replied:
Regarding to maxFrameRate: should it be also applied in case the resolution is lower than specified limit? For example limit is: 1920x1080@30 (maxFrameRate = 30) and supportsType is for lower resolution with higher frameRate (ex.: resolution=720×480, frameRate: 60)?

eocanha replied:
Applying the frame rate limit only if the resolution is the maximum one makes sense, yes.

I don't know how to treat corner cases like if the webpage is asking for 1900x1070@60, or 720x480@200. I guess we can take care of those in the future if they become a problem.

mcatanzaro commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -357,7 +357,7 @@ SET_AND_EXPOSE_TO_BUILD(USE_NICOSIA TRUE)
 SET_AND_EXPOSE_TO_BUILD(HAVE_OS_DARK_MODE_SUPPORT 1)
 
-SET_AND_EXPOSE_TO_BUILD(USE_FULLHD_VIDEO_DECODING_LIMIT ${ENABLE_FULLHD_VIDEO_DECODING_LIMIT})
+add_definitions(-DVIDEO_DECODING_LIMIT="${VIDEO_DECODING_LIMIT}")
```
I'm going to suggest changing this from a build option to a runtime API so that you don't need a new build to change the limit.

That said, if you do prefer to stick with a build option, I would define it only for the one file that actually needs it, MediaPlayerPrivateGStreamerMSE.cpp, rather than here. What you have currently is going to require a full rebuild of all of WebKit each time you want to change the limit.

eocanha replied:
IMHO, for the typical use case where we need this feature (to adapt WPE to the capabilities of different devices, typically set-top-boxes) I think there's no advantage on being able to change the limit at runtime and a build option should be enough.

What would be the way to define it for a single file in cmake, as you suggest?

mcatanzaro replied:
You could try this in Source/WebCore/platform/GStreamer.cmake:

```
set_source_files_properties(MediaPlayerPrivateGStreamerMSE.cpp PROPERTIES COMPILE_DEFINITIONS "-DVIDEO_DECODING_LIMIT=${VIDEO_DECODING_LIMIT}")
```

(Note this will work only because it is not unified: you won't be able to add it to unified build in the future.)

> IMHO, for the typical use case where we need this feature (to adapt WPE to the capabilities of different devices, typically set-top-boxes) I think there's no advantage on being able to change the limit at runtime and a build option should be enough.

OK then, though it means you won't be able to enable the video decoding limit in any standard builds, e.g. WebKit flatpak runtime builds, or Linux distro builds. Perhaps a bit disappointing if you were hoping to use the WebKit flatpak runtime on Raspberry Pi, for example?

mcatanzaro commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
```diff
@@ -116,6 +116,11 @@ class MediaPlayerPrivateGStreamerMSE : public MediaPlayerPrivateGStreamer {
+    static unsigned mediaMaxWidth;
+    static unsigned mediaMaxHeight;
+    static unsigned mediaMaxFrameRate;
+    static bool isVideoDecoderLimitDefaultParsed;
```
Minor improvement: you could declare these inside `MediaPlayerPrivateGStreamerMSE::supportsType` instead, to give them the smallest scope necessary and avoid the need to initialize them in a different place from where they are declared.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -376,6 +419,18 @@ void MediaPlayerPrivateGStreamerMSE::getSupportedTypes(HashSet<String, ASCIICase
+    static unsigned mediaMaxHeight = 0;
+    static unsigned mediaMaxFrameRate = 0;
+#ifdef VIDEO_DECODING_LIMIT
+    static bool isVideoDecoderLimitDefaultParsed = false;
```
To avoid this, please use `std::once` and `std::once_flag`. Now that I see this code I think you can forget about using `std::optional<std::tuple<...>>`. I think the proper thing here would be capturing `&mediaMaxWidth, &mediaMaxHeight, &mediaMaxFrameRage` in the lamba that is going to be passed to the `std::once` call and inside that there you do the parsing.

I guess you also need to set a default somewhere and I think the proper place would be to do it in the declaration of the static variables.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -376,6 +419,18 @@ void MediaPlayerPrivateGStreamerMSE::getSupportedTypes(HashSet<String, ASCIICase
+    if (!isVideoDecoderLimitDefaultParsed) {
+        isVideoDecoderLimitDefaultParsed = true;
+        if (!videoDecoderLimitsDefaults(mediaMaxWidth, mediaMaxHeight, mediaMaxFrameRate))
+            GST_ERROR("Parsing VIDEO_DECODING_LIMIT failed");
```
Once you set defaults, you can just `GST_WARNING` here because things are going to "work" but maybe not as expected. A warning should be enough IMHO but also, please call `ASSERT_NOT_REACHED()` in that case so we have a partially working thing in release but a crash in debug.

pgorszkowski-igalia replied:
If I use ASSERT_NOT_REACHED() I have style check error:

`ERROR: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:429:  ASSERT_NOT_REACHED() statement fallthrough may result in unexpected code execution.  [security/assertion_fallthrough] [4]`

Any idea how to fix it?

aperezdc replied:
This is because `ASSERT_NOT_REACHED()` expands to nothing in release builds. Try `RELEASE_ASSERT_NOT_REACHED()` instead, which will always work regardless of the build type.

pgorszkowski-igalia replied:
I think that(expand to nothing in release) is what we want. @calvaris : do you think I should use RELEASE_ASSERT_NOT_REACHED() here?

calvaris replied:
No, definitely do not use `RELEASE_ASSERT_NOT_REACHED()`, that will crash in release mode. We just want to crash in debug mode.

calvaris reviewed (commented):
Other than the style issue that I think it is something we have to live with, I miss the defaults for the values so that we don't have to check for default values. You didn't comment anything about this.

pgorszkowski-igalia committed:
[GStreamer][MSE][CMake] make configurable video decoder limits
https://bugs.webkit.org/show_bug.cgi?id=245660

Reviewed by Xabier Rodriguez-Calvar.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(videoDecoderLimitsDefaults):
(WebCore::MediaPlayerPrivateGStreamerMSE::MediaPlayerPrivateGStreamerMSE):
(WebCore::MediaPlayerPrivateGStreamerMSE::supportsType):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/255380@main
